Ignore protected member functions

## Expected Behavior
no error

## Actual Behavior
has error

## Steps to Reproduce the Problem

```
class Hello {
public:
    Hello();
    int public_function();
protected:
    int protected_function();
};
```

with

```
include_cpp! {
    #include "input.h"
    generate!("Hello")
}
```

yields

```
 running: "c++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-arch" "x86_64" "-I" "src" "-I" "/Users/tom/QAPMain/Development/frontend/ncnn_rust/target/debug/build/ncnn_rust-2cef91cead1f1e2d/out/autocxx-build-dir/include" "-Wall" "-Wextra" "-DMY_BUILD_TARGET_MACOS=1" "-std=c++14" "-o" "/Users/tom/QAPMain/Development/frontend/ncnn_rust/target/debug/build/ncnn_rust-2cef91cead1f1e2d/out/autocxx-build-dir/cxx/gen0.o" "-c" "/Users/tom/QAPMain/Development/frontend/ncnn_rust/target/debug/build/ncnn_rust-2cef91cead1f1e2d/out/autocxx-build-dir/cxx/gen0.cxx"
  cargo:warning=/Users/tom/QAPMain/Development/frontend/ncnn_rust/target/debug/build/ncnn_rust-2cef91cead1f1e2d/out/autocxx-build-dir/cxx/gen0.cxx:149:57: error: 'protected_function' is a protected member of 'Hello'
  cargo:warning=  ::c_int (::Hello::*protected_function$)() = &::Hello::protected_function;
  cargo:warning=                                                        ^
  cargo:warning=src/input.h:30:9: note: declared protected here
  cargo:warning=    int protected_function();
  cargo:warning=        ^
  cargo:warning=1 error generated.
  exit status: 1

  --- stderr
  Start ncnn_rust build script target=Macos


  error occurred: Command "c++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-arch" "x86_64" "-I" "src" "-I" "/Users/tom/QAPMain/Development/frontend/ncnn_rust/target/debug/build/ncnn_rust-2cef91cead1f1e2d/out/autocxx-build-dir/include" "-Wall" "-Wextra" "-DMY_BUILD_TARGET_MACOS=1" "-std=c++14" "-o" "/Users/tom/QAPMain/Development/frontend/ncnn_rust/target/debug/build/ncnn_rust-2cef91cead1f1e2d/out/autocxx-build-dir/cxx/gen0.o" "-c" "/Users/tom/QAPMain/Development/frontend/ncnn_rust/target/debug/build/ncnn_rust-2cef91cead1f1e2d/out/autocxx-build-dir/cxx/gen0.cxx" with args "c++" did not execute successfully (status code exit status: 1).
```

Thus, is there, at least, some workarounds? I have tried `block!` on `Hello::protected_function` but completely no use. This make autocxx not useable...

## Specifications

- Version: latest
- Platform: macos

--- a/engine/src/conversion/analysis/fun/mod.rs
+++ b/engine/src/conversion/analysis/fun/mod.rs
@@ -24,7 +24,7 @@ use crate::{
             fun::function_wrapper::CppFunctionKind,
             type_converter::{self, add_analysis, TypeConversionContext, TypeConverter},
         },
-        api::{ApiName, FuncToConvert, SubclassName},
+        api::{ApiName, CppVisibility, FuncToConvert, SubclassName},
         convert_error::ConvertErrorWithContext,
         convert_error::ErrorContext,
         error_reporter::{convert_apis, report_any_error},
@@ -545,9 +545,9 @@ impl<'a> FnAnalyzer<'a> {
             )
         };
 
-        // Skip private methods; but if we've a private constructor, keep
+        // Skip private & protected methods; but if we've a private constructor, keep
         // a note of it.
-        if fun.is_private {
+        if fun.cpp_vis != CppVisibility::Public {
             if let FnKind::Method(self_ty, MethodKind::Constructor) = &kind {
                 self.has_unrepresentable_constructors
                     .insert(self_ty.clone());
@@ -1020,7 +1020,7 @@ impl<'a> FnAnalyzer<'a> {
                         output: ReturnType::Default,
                         vis: parse_quote! { pub },
                         is_pure_virtual: false,
-                        is_private: false,
+                        cpp_vis: CppVisibility::Public,
                         is_move_constructor: false,
                         unused_template_param: false,
                         return_type_is_reference: false,

--- a/engine/src/conversion/analysis/fun/subclass.rs
+++ b/engine/src/conversion/analysis/fun/subclass.rs
@@ -18,7 +18,7 @@ use syn::{parse_quote, FnArg, PatType, Type, TypePtr};
 
 use crate::conversion::analysis::fun::ReceiverMutability;
 use crate::conversion::analysis::pod::PodPhase;
-use crate::conversion::api::{FuncToConvert, RustSubclassFnDetails, SubclassName};
+use crate::conversion::api::{CppVisibility, FuncToConvert, RustSubclassFnDetails, SubclassName};
 use crate::{
     conversion::{
         analysis::fun::function_wrapper::{
@@ -62,7 +62,7 @@ pub(super) fn create_subclass_fn_wrapper(
         output: fun.output.clone(),
         vis: fun.vis.clone(),
         is_pure_virtual: false,
-        is_private: false,
+        cpp_vis: CppVisibility::Public,
         is_move_constructor: false,
         unused_template_param: fun.unused_template_param,
         original_name: None,
@@ -168,7 +168,7 @@ pub(super) fn create_subclass_constructor_wrapper(
         output: fun.output.clone(),
         vis: fun.vis.clone(),
         is_pure_virtual: false,
-        is_private: fun.is_private,
+        cpp_vis: fun.cpp_vis,
         is_move_constructor: false,
         original_name: None,
         unused_template_param: fun.unused_template_param,

--- a/engine/src/conversion/api.rs
+++ b/engine/src/conversion/api.rs
@@ -48,6 +48,14 @@ pub(crate) struct ImplBlockDetails {
     pub(crate) ty: Ident,
 }
 
+/// C++ visibility.
+#[derive(Clone, PartialEq, Eq, Copy)]
+pub(crate) enum CppVisibility {
+    Public,
+    Protected,
+    Private,
+}
+
 /// A C++ function for which we need to generate bindings, but haven't
 /// yet analyzed in depth. This is little more than a `ForeignItemFn`
 /// broken down into its constituent parts, plus some metadata from the
@@ -60,7 +68,7 @@ pub(crate) struct FuncToConvert {
     pub(crate) output: ReturnType,
     pub(crate) vis: Visibility,
     pub(crate) is_pure_virtual: bool,
-    pub(crate) is_private: bool,
+    pub(crate) cpp_vis: CppVisibility,
     pub(crate) is_move_constructor: bool,
     pub(crate) unused_template_param: bool,
     pub(crate) return_type_is_reference: bool,

--- a/engine/src/conversion/parse/parse_foreign_mod.rs
+++ b/engine/src/conversion/parse/parse_foreign_mod.rs
@@ -12,7 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use crate::conversion::api::ApiName;
+use crate::conversion::api::{ApiName, CppVisibility};
 use crate::conversion::doc_attr::get_doc_attr;
 use crate::conversion::error_reporter::report_any_error;
 use crate::conversion::{
@@ -76,6 +76,16 @@ impl ParseForeignMod {
         self.ignored_apis.append(&mut extra_apis);
     }
 
+    fn get_cpp_visibility(item: &ForeignItemFn) -> CppVisibility {
+        if Self::has_attr(&item, "bindgen_visibility_private") {
+            CppVisibility::Private
+        } else if Self::has_attr(&item, "bindgen_visibility_protected") {
+            CppVisibility::Protected
+        } else {
+            CppVisibility::Public
+        }
+    }
+
     fn parse_foreign_item(
         &mut self,
         i: ForeignItem,
@@ -83,7 +93,7 @@ impl ParseForeignMod {
     ) -> Result<(), ConvertErrorWithContext> {
         match i {
             ForeignItem::Fn(item) => {
-                let is_private = Self::has_attr(&item, "bindgen_visibility_private");
+                let cpp_vis = Self::get_cpp_visibility(&item);
                 let is_pure_virtual = Self::has_attr(&item, "bindgen_pure_virtual");
                 let unused_template_param =
                     Self::has_attr(&item, "bindgen_unused_template_param_in_arg_or_return");
@@ -101,7 +111,7 @@ impl ParseForeignMod {
                     output: item.sig.output,
                     vis: item.vis,
                     is_pure_virtual,
-                    is_private,
+                    cpp_vis,
                     is_move_constructor,
                     unused_template_param,
                     return_type_is_reference,

--- a/integration-tests/src/tests.rs
+++ b/integration-tests/src/tests.rs
@@ -7040,7 +7040,7 @@ fn test_abstract_up() {
 }
 
 #[test]
-fn test_protected_method() {
+fn test_class_having_protected_method() {
     let hdr = indoc! {"
         class A {
         protected:
@@ -7052,7 +7052,7 @@ fn test_protected_method() {
 }
 
 #[test]
-fn test_private_method() {
+fn test_class_having_private_method() {
     let hdr = indoc! {"
         class A {
         private:

--- a/integration-tests/src/tests.rs
+++ b/integration-tests/src/tests.rs
@@ -7042,10 +7042,11 @@ fn test_abstract_up() {
 #[test]
 fn test_class_having_protected_method() {
     let hdr = indoc! {"
-        class A {
-        protected:
-            uint32_t protected_method();
-        };
+    #include <cstdint>
+    class A {
+    protected:
+        inline uint32_t protected_method() { return 0; }
+    };
     "};
     let rs = quote! {};
     run_test("", hdr, rs, &[], &["A"]);
@@ -7054,10 +7055,11 @@ fn test_class_having_protected_method() {
 #[test]
 fn test_class_having_private_method() {
     let hdr = indoc! {"
-        class A {
-        private:
-            uint32_t private_method();
-        };
+    #include <cstdint>
+    class A {
+    private:
+        inline uint32_t private_method() { return 0; }
+    };
     "};
     let rs = quote! {};
     run_test("", hdr, rs, &[], &["A"]);

--- a/integration-tests/src/tests.rs
+++ b/integration-tests/src/tests.rs
@@ -7039,6 +7039,30 @@ fn test_abstract_up() {
     run_test("", hdr, rs, &["A", "get_a"], &[]);
 }
 
+#[test]
+fn test_protected_method() {
+    let hdr = indoc! {"
+        class A {
+        protected:
+            uint32_t protected_method();
+        };
+    "};
+    let rs = quote! {};
+    run_test("", hdr, rs, &[], &["A"]);
+}
+
+#[test]
+fn test_private_method() {
+    let hdr = indoc! {"
+        class A {
+        private:
+            uint32_t private_method();
+        };
+    "};
+    let rs = quote! {};
+    run_test("", hdr, rs, &[], &["A"]);
+}
+
 // Yet to test:
 // - Ifdef
 // - Out param pointers

